reducer/src/main/java/com/graphicsfuzz/reducer/reductionopportunities/RemoveRedundantUniformMetadataReductionOpportunities.java

```diff
@@ -0,0 +1,94 @@
+ * limitations under the License.
+ */
+
+package com.graphicsfuzz.reducer.reductionopportunities;
+
+import com.graphicsfuzz.common.ast.TranslationUnit;
+import com.graphicsfuzz.common.ast.decl.VariableDeclInfo;
+import com.graphicsfuzz.common.ast.decl.VariablesDeclaration;
+import com.graphicsfuzz.common.ast.expr.VariableIdentifierExpr;
+import com.graphicsfuzz.common.transformreduce.ShaderJob;
+import com.graphicsfuzz.common.util.ListConcat;
+import com.graphicsfuzz.common.util.PipelineInfo;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class RemoveRedundantUniformMetadataReductionOpportunities
+    extends ReductionOpportunitiesBase<RemoveRedundantUniformMetadataReductionOpportunity> {
+
+  private final PipelineInfo pipelineInfo;
+  private List<String> redundantUniformNames;
+
+  public RemoveRedundantUniformMetadataReductionOpportunities(TranslationUnit tu,
+                                                              ReducerContext context,
+                                                              PipelineInfo pipelineInfo) {
+    super(tu, context);
+    this.pipelineInfo = pipelineInfo;
+  }
+
+  static List<RemoveRedundantUniformMetadataReductionOpportunity> findOpportunities(
+      ShaderJob shaderJob,
+      ReducerContext context) {
+    return shaderJob.getShaders()
+        .stream()
+        .map(item -> findOpportunitiesForShader(item, context, shaderJob.getPipelineInfo()))
+        .reduce(Arrays.asList(), ListConcat::concatenate);
+  }
+
+  @Override
+  public void visitTranslationUnit(TranslationUnit translationUnit) {
+    redundantUniformNames = new ArrayList<>(pipelineInfo.getUniformNames());
+    super.visitTranslationUnit(translationUnit);
+
+    for (String uniformName : redundantUniformNames) {
+      addOpportunity(new RemoveRedundantUniformMetadataReductionOpportunity(uniformName,
+          pipelineInfo, getVistitationDepth()));
+    }
+  }
+
+  private static List<RemoveRedundantUniformMetadataReductionOpportunity>
+      findOpportunitiesForShader(TranslationUnit tu,
+                               ReducerContext context,
+                               PipelineInfo pipelineInfo) {
+    RemoveRedundantUniformMetadataReductionOpportunities finder =
+          new RemoveRedundantUniformMetadataReductionOpportunities(tu, context, pipelineInfo);
+    finder.visit(tu);
+    return finder.getOpportunities();
+  }
+
+  @Override
+  public void visitVariablesDeclaration(VariablesDeclaration variablesDeclaration) {
+    // This prevents removing a uniform declared but not used.
+    super.visitVariablesDeclaration(variablesDeclaration);
+    for (VariableDeclInfo vdi : variablesDeclaration.getDeclInfos()) {
+      removeRedundantUniformName(vdi.getName());
+    }
+  }
+
+  @Override
+  public void visitVariableIdentifierExpr(VariableIdentifierExpr variableIdentifierExpr) {
+    super.visitVariableIdentifierExpr(variableIdentifierExpr);
```

Comment:
This should not be needed.  We cannot use a variable if we have not declared it, and if we tag uniforms as being needed if they are declared then we are fine.

Reply:
Fixed